File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/05．治疗护盾联动.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__Delete = ____lualib.__TS__Delete
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____require_result_0 = require("系统.04．伤害系统.02．治疗系统.01．核心功能")
local registerAppliedFinalHealListener = ____require_result_0.registerAppliedFinalHealListener
local _____6CBB_7597_62A4_76FE_8054_52A8_8868 = {}
local _____6CBB_7597_62A4_76FE_8054_52A8_8BA1_6570 = 0
local _____6CBB_7597_62A4_76FE_8054_52A8_5B9E_73B0 = __TS__Class()
_____6CBB_7597_62A4_76FE_8054_52A8_5B9E_73B0.name = "治疗护盾联动实现"
function _____6CBB_7597_62A4_76FE_8054_52A8_5B9E_73B0.prototype.____constructor(self, _____540D_79F0, _____53C2_6570)
    self["已停止"] = false
    self["名称"] = _____540D_79F0
    self["参数"] = _____53C2_6570
    _____6CBB_7597_62A4_76FE_8054_52A8_8BA1_6570 = _____6CBB_7597_62A4_76FE_8054_52A8_8BA1_6570 + 1
    self["控制器ID"] = _____6CBB_7597_62A4_76FE_8054_52A8_8BA1_6570
    _____6CBB_7597_62A4_76FE_8054_52A8_8868[self["控制器ID"]] = self
end
_____6CBB_7597_62A4_76FE_8054_52A8_5B9E_73B0.prototype["处理治疗"] = function(self, event)
    if self["已停止"] or self["参数"]["on治疗"] == nil or not self["匹配单位"](self, event) then
        return
    end
    if self["参数"]["过滤事件"] ~= nil and not self["参数"]["过滤事件"](event) then
        return
    end
    self["参数"]["on治疗"](event)
end
_____6CBB_7597_62A4_76FE_8054_52A8_5B9E_73B0.prototype["处理护盾"] = function(self, event)
    if self["已停止"] or self["参数"]["on护盾"] == nil or not self["匹配单位"](self, event) then
        return
    end
    if self["参数"]["过滤事件"] ~= nil and not self["参数"]["过滤事件"](event) then
        return
    end
    self["参数"]["on护盾"](event)
end
_____6CBB_7597_62A4_76FE_8054_52A8_5B9E_73B0.prototype["停止"] = function(self)
    if self["已停止"] then
        return
    end
    self["已停止"] = true
    __TS__Delete(_____6CBB_7597_62A4_76FE_8054_52A8_8868, self["控制器ID"])
end
_____6CBB_7597_62A4_76FE_8054_52A8_5B9E_73B0.prototype["匹配单位"] = function(self, event)
    if self["参数"]["单位"] == nil then
        return true
    end
    local _____65B9_5411 = self["参数"]["监听方向"] or "双向"
    if _____65B9_5411 == "自己获得" then
        return event["目标单位"] == self["参数"]["单位"]
    end
    if _____65B9_5411 == "自己给予" then
        return event["来源单位"] == self["参数"]["单位"]
    end
    return event["目标单位"] == self["参数"]["单位"] or event["来源单位"] == self["参数"]["单位"]
end
____exports["创建治疗护盾联动"] = function(_____53C2_6570)
    return __TS__New(_____6CBB_7597_62A4_76FE_8054_52A8_5B9E_73B0, _____53C2_6570["名称"] or "治疗护盾联动", _____53C2_6570)
end
____exports["通知获得护盾事件"] = function(_____6765_6E90_5355_4F4D, _____76EE_6807_5355_4F4D, _____6570_503C, _____6807_7B7E, _____539F_59CB_53C2_6570)
    local event = {
        ["来源单位"] = _____6765_6E90_5355_4F4D,
        ["目标单位"] = _____76EE_6807_5355_4F4D,
        ["数值"] = _____6570_503C,
        ["标签"] = _____6807_7B7E,
        ["原始参数"] = _____539F_59CB_53C2_6570
    }
    for key in pairs(_____6CBB_7597_62A4_76FE_8054_52A8_8868) do
        local _____63A7_5236_5668 = _____6CBB_7597_62A4_76FE_8054_52A8_8868[key]
        if _____63A7_5236_5668 ~= nil then
            _____63A7_5236_5668["处理护盾"](_____63A7_5236_5668, event)
        end
    end
end
local function ____on_6700_7EC8_6CBB_7597_8054_52A8(source, target, amount, isItemHeal)
    if amount <= 0 then
        return
    end
    local event = {["来源单位"] = source, ["目标单位"] = target, ["数值"] = amount, ["是否物品治疗"] = isItemHeal}
    for key in pairs(_____6CBB_7597_62A4_76FE_8054_52A8_8868) do
        local _____63A7_5236_5668 = _____6CBB_7597_62A4_76FE_8054_52A8_8868[key]
        if _____63A7_5236_5668 ~= nil then
            _____63A7_5236_5668["处理治疗"](_____63A7_5236_5668, event)
        end
    end
end
registerAppliedFinalHealListener(____on_6700_7EC8_6CBB_7597_8054_52A8)
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/05．治疗护盾联动.ts
/** @noSelfInFile */

const { registerBeforeAppliedFinalHealListener, registerAppliedFinalHealListener } = require("系统.04．伤害系统.02．治疗系统.01．核心功能") as {
  registerBeforeAppliedFinalHealListener: (this: void, cb: (this: void, source: any, target: any, amount: number, isItemHeal: boolean) => void) => void;
  registerAppliedFinalHealListener: (this: void, cb: (this: void, source: any, target: any, amount: number, isItemHeal: boolean) => void) => void;
};

export interface 治疗护盾联动事件 {
  来源单位: any;
  目标单位: any;
  数值: number;
  是否物品治疗?: boolean;
  标签?: string;
  原始参数?: any;
}

export interface 治疗护盾联动参数 {
  名称?: string;
  单位?: any;
  监听方向?: "自己获得" | "自己给予" | "双向";
  治疗触发阶段?: "治疗开始" | "治疗完成";
  过滤事件?: (this: void, event: 治疗护盾联动事件) => boolean;
  on治疗?: (this: void, event: 治疗护盾联动事件) => void;
  on护盾?: (this: void, event: 治疗护盾联动事件) => void;
}

export interface 治疗护盾联动控制器 {
  readonly 名称: string;
  停止(): void;
}

const 治疗护盾联动表: Record<number, 治疗护盾联动实现> = {};
let 治疗护盾联动计数 = 0;

class 治疗护盾联动实现 implements 治疗护盾联动控制器 {
  readonly 名称: string;
  readonly 控制器ID: number;
  private 参数: 治疗护盾联动参数;
  private 已停止 = false;

  constructor(名称: string, 参数: 治疗护盾联动参数) {
    this.名称 = 名称;
    this.参数 = 参数;
    this.控制器ID = ++治疗护盾联动计数;
    治疗护盾联动表[this.控制器ID] = this;
  }

  处理治疗(event: 治疗护盾联动事件, 阶段: "治疗开始" | "治疗完成"): void {
    if (this.已停止 || this.参数.on治疗 == null || !this.匹配单位(event)) return;
    if ((this.参数.治疗触发阶段 ?? "治疗完成") !== 阶段) return;
    if (this.参数.过滤事件 != null && !this.参数.过滤事件(event)) return;
    this.参数.on治疗(event);
  }

  处理护盾(event: 治疗护盾联动事件): void {
    if (this.已停止 || this.参数.on护盾 == null || !this.匹配单位(event)) return;
    if (this.参数.过滤事件 != null && !this.参数.过滤事件(event)) return;
    this.参数.on护盾(event);
  }

  停止(): void {
    if (this.已停止) return;
    this.已停止 = true;
    delete 治疗护盾联动表[this.控制器ID];
  }

  private 匹配单位(event: 治疗护盾联动事件): boolean {
    if (this.参数.单位 == null) return true;
    const 方向 = this.参数.监听方向 ?? "双向";
    if (方向 === "自己获得") return event.目标单位 === this.参数.单位;
    if (方向 === "自己给予") return event.来源单位 === this.参数.单位;
    return event.目标单位 === this.参数.单位 || event.来源单位 === this.参数.单位;
  }
}

export function 创建治疗护盾联动(this: void, 参数: 治疗护盾联动参数): 治疗护盾联动控制器 {
  return new 治疗护盾联动实现(参数.名称 ?? "治疗护盾联动", 参数);
}

export function 通知获得护盾事件(this: void, 来源单位: any, 目标单位: any, 数值: number, 标签?: string, 原始参数?: any): void {
  const event: 治疗护盾联动事件 = { 来源单位, 目标单位, 数值, 标签, 原始参数 };
  for (const key in 治疗护盾联动表) {
    const 控制器 = 治疗护盾联动表[key];
    if (控制器 != null) 控制器.处理护盾(event);
  }
}

function 分发治疗联动(this: void, source: any, target: any, amount: number, isItemHeal: boolean, 阶段: "治疗开始" | "治疗完成"): void {
  if (amount <= 0) return;
  const event: 治疗护盾联动事件 = { 来源单位: source, 目标单位: target, 数值: amount, 是否物品治疗: isItemHeal };
  for (const key in 治疗护盾联动表) {
    const 控制器 = 治疗护盾联动表[key];
    if (控制器 != null) 控制器.处理治疗(event, 阶段);
  }
}

function on治疗开始联动(this: void, source: any, target: any, amount: number, isItemHeal: boolean): void {
  分发治疗联动(source, target, amount, isItemHeal, "治疗开始");
}

function on最终治疗联动(this: void, source: any, target: any, amount: number, isItemHeal: boolean): void {
  分发治疗联动(source, target, amount, isItemHeal, "治疗完成");
}

registerBeforeAppliedFinalHealListener(on治疗开始联动);
registerAppliedFinalHealListener(on最终治疗联动);
